packages/wallet/ui/src/bridge-dapp.jsx

```diff
@@ -0,0 +1,180 @@
+
+/**
+ * Sends a message to the dapp in the parent window.
+ *
+ * @param {*} payload
+ */
+const sendMessage = payload => window.parent.postMessage(payload, '*');
+
+/**
+ * Ensures this bridge is being loaded inside another window.
+ */
+const checkParentWindow = () => {
+  const me = window;
+  const { parent } = window;
+  if (me === parent) {
+    throw Error('window.parent === parent!!!');
+  }
+};
+
+/**
+ * Requests a dapp at a given origin to connect to the wallet UI. This will
+ * prompt the user to accept the dapp in the wallet and allow it to propose
+ * offers if accepted.
+ *
+ * @param {string} origin
+ * @param {string} chainId
+ * @param {string} address
+ * @param {string} proposedPetname
+ */
+const requestDappConnection = (origin, chainId, address, proposedPetname) => {
+  const dapp = loadDapp(chainId, address, origin);
+  if (dapp) {
+    return;
+  }
+  upsertDapp(chainId, address, { origin, petname: proposedPetname });
+};
+
+/**
+ * Watches for changes in local storage to the dapp's approval status and
+ * notifies the dapp.
+ *
+ * @param {string} origin
+ * @param {string} chainId
+ * @param {string} address
+ * @param {boolean} latestValue
+ */
+const watchDappApproval = (origin, chainId, address, latestValue) => {
+  watchDapps(chainId, address, dapps => {
+    console.log('something happened!!!', dapps);
+    const dapp = dapps.find(d => d.origin === origin);
+    const isDappApproved = dapp && dapp.isEnabled;
+    if (isDappApproved !== latestValue) {
+      sendMessage({
+        type: BridgeProtocol.dappApprovalChanged,
+        isDappApproved,
+      });
+      latestValue = isDappApproved;
+    }
+  });
+};
+
+const createAndAddOffer = (origin, chainId, address, config) => {
+  const currentTime = new Date().getTime();
+  const id = currentTime;
+
+  const offer = {
+    id,
+    instancePetname: `instance@${config.instanceHandleBoardId}`,
+    requestContext: { origin },
+    meta: {
+      id,
+      creationStamp: currentTime,
+    },
+    status: 'proposed',
+    ...config,
+  };
+  addOffer(chainId, address, offer);
+};
+
+/**
+ * Notifies the dapp about whether it's approved or not and watches for changes
+ * to its approval status.
+ *
+ * @param {string} origin
+ * @param {string} chainId
+ * @param {string} address
+ */
+const checkIfDappApproved = (origin, chainId, address) => {
+  const dapp = loadDapp(chainId, address, origin);
+  const isDappApproved = dapp && dapp.isEnabled;
+  sendMessage({
+    type: BridgeProtocol.checkIfDappApproved,
+    isDappApproved,
+  });
+  watchDappApproval(origin, chainId, address, isDappApproved);
+};
+
+const handleIncomingMessages = () => {
+  /** @type { string } */
+  let origin;
+  /** @type { string } */
+  let address;
+  /** @type { string } */
+  let chainId;
+
+  window.addEventListener('message', ev => {
+    const type = ev.data?.type;
+    if (!type?.startsWith('agoric_')) {
+      return;
+    }
+    if (origin === undefined) {
```

Comment:
surprising that whatever the first message is gets to define the origin, which then remains fixed. seems ripe for race conditions. can `origin` be a parameter to `handleIncomingMessages`?